pkg/upgrade/upgrade_state.go

```
@@ -0,0 +1,226 @@
+}
+
+func NewClusterUpgradeState() ClusterUpgradeState {
+	return ClusterUpgradeState{NodeStates: make(map[string][]*NodeUpgradeState)}
+}
+
+type ClusterUpgradeStateManager struct {
+	K8sClient         client.Client
+	Log               logr.Logger
+	DrainManager      DrainManager
+	PodRestartManager PodRestartManager
+}
+
+func NewClusterUpdateStateManager(
+	drainManager DrainManager,
+	podRestartManager PodRestartManager,
+	log logr.Logger,
+	k8sClient client.Client) *ClusterUpgradeStateManager {
+	manager := &ClusterUpgradeStateManager{
+		DrainManager:      drainManager,
+		PodRestartManager: podRestartManager,
+		Log:               log,
+		K8sClient:         k8sClient,
+	}
+
+	return manager
+}
+
+func (m *ClusterUpgradeStateManager) ApplyState(ctx context.Context, currentState *ClusterUpgradeState) {
+	m.Log.V(consts.LogLevelInfo).Info("State Manager, got state update")
+
+	policy := currentState.Policy
+	if !policy.AutoUpgrade {
+		m.Log.V(consts.LogLevelInfo).Info("Driver auto upgrade is disabled, skipping")
+		return
+	}
+
+	m.Log.V(consts.LogLevelInfo).Info("Node states:",
+		"Unknown", len(currentState.NodeStates[Unknown]),
+		Ready, len(currentState.NodeStates[Ready]),
+		UpgradeRequired, len(currentState.NodeStates[UpgradeRequired]),
+		Drain, len(currentState.NodeStates[Drain]),
+		DrainFailed, len(currentState.NodeStates[DrainFailed]),
+		PodRestart, len(currentState.NodeStates[PodRestart]))
+
+	upgradesInProgress := len(currentState.NodeStates[Drain]) +
+		len(currentState.NodeStates[PodRestart]) +
+		len(currentState.NodeStates[DrainFailed])
+
+	upgradesAvailable := policy.MaxParallelUpgrades - upgradesInProgress
+
+	m.Log.V(consts.LogLevelInfo).Info("Upgrades in progress",
+		"currently in progress", upgradesInProgress,
+		"max parallel upgrades", policy.MaxParallelUpgrades,
+		"upgrade slots available", upgradesAvailable)
+
+	// First, check if unknown or ready nodes need to be upgraded
+	m.ProcessReadyOrUnknownNodes(currentState, Unknown)
```

Please, do not ignore errors in all Process handlers. I thinks we should rise all critical errors up to Reconcile loop and rely on controller-runtime error handling logic which will schedule retry for us.